source/program/remote_api.cpp

```diff
@@ -54,36 +76,77 @@ void PrepareThread() {
 
     int rval = nn::socket::Bind(g_TcpSocket, (struct sockaddr*)&serverAddr, sizeof(serverAddr));
     if (rval < 0) EXL_ABORT(69);
-    
+
     rval = nn::socket::Listen(g_TcpSocket, 1);
     if (rval < 0) EXL_ABORT(69);
 }
 
-void SocketSpawn(void*) {
+void AddPacketToSendBuffer(PacketBuffer& buffer) {
+    nn::os::LockMutex(&sendBufferLock);
+    sendBuffer.push_back(std::move(buffer));
+    nn::os::UnlockMutex(&sendBufferLock);
+}
+
+void ReceiveLogic() {
+    // don't recv new packets while we waiting for game loop otherwise we need to make a copy of our receive buffer
+    if (readyForGameThread.load()) return;
```

Comment:
@henriquegemignani 
Should we decrement the `keepAlive` only if it passes this check?
Otherwise it's "luck based" if you lose the connection.
E.g.: Receive an item right the moment you hit a loading screen => `readyForGameThread` will be true for all upcoming keep alive packets and therefore are not processed until the game is ready.

Reply:
What is the purpose of the keep alive? Why does it have to be 2000ms?

Reply:
The keep alive is just an arbitrary number.
I implemented it because, I haven't found a guaranteed (ryujinx) way of determining if the connection is alive or not.
If keepAlive is 0 exlaunch goes back to the outer loop to accept a new connection and you can reconnect from randovania.
I just made it that randovania sends it every 2 seconds and if exlaunch did not received one in 10 seconds, you must reconnect.

Reply:
When I tested the old version, Randovania's polling would be interrupted by over 30s during loading screens. So I imagine either the keep alive need to be that long, or consider it acceptable to disconnect after a load.

Reply:
The keep alive can be processed while on a loading screen. The only thing that cannot be processed are lua executions. They are "queued" until loading is done and therefore the reception of packets is blocked until the lua execution is finished. (We only have one receive buffer).
And that's the edge case: Lua execution (receiving an item) during a loading screen which takes longer than 10 seconds.

As I'm writing this...yeah just ignore the keep alive decrement as long as readyForGameThread is true should do the job.

Reply:
Keep alive is now not decremented when receiving is on hold, which is the case when we are waiting for the game to execute the lua code in the receive buffer.